lib/kcerts/cache.go

```diff
@@ -36,19 +38,27 @@ func FetchSSHAgentFromCache(store cache.Store) (*SSHAgent, error) {
 
 func WriteAgentToCache(store cache.Store, agent *SSHAgent) error {
 	sshEnkitCache, _, err := store.Get(SSHCacheKey)
+
+	killFunc := func() {
```

Comment:
no, do the opposite here: if writing succeeds, than set the "Close" to a noop function. Otherwise, leave it as is.
In the function that creates an agent, set the "Close" to "Kill" the agent.

Rationale:
As the code is written now, if one uses the API in different ways, it can lead to buggy / wrong results.

For example, let's say I change the code (or use the functions) so that `WriteAgentToCache` is always invoked, even when I've loaded the agent from the environment, or when I've read it from cache. If `WriteAgentToCache` fails, we end up killing *innocent* agents!! Some poor agents that don't belong to us.
(from documentation, this use of the functions is perfectly legitimate).

Another example: I create an agent invoking the function below. Then I never write to cache. I call Close() assuming the agent will be terminated, but it won't! As it's WriteAgentToCache that kills the agent in case of error.

Basically:
- the function that creates the SSHAgent owns the agent, and knows how to kill it.
- WriteAgentToCache, if it succeeds, should prevent the agent from being killed, as it's now part of the cache.
if it fails, it should terminate the agent as whoever owns it intended for it to be terminated.